test/system/kubernetes/fluentBitRotater.sh

```diff
@@ -0,0 +1,261 @@
+    fi
+}
+
+items() {
+    local spaces=$(echo "$*" | head -n 1 | tr -cd ' \t' | wc -c)
+    if [ -z "$*" ]; then
+        echo "$spaces"
+    else
+        echo "$((spaces+1))"
+    fi
+}
+
+used_kib() {
+    du -s $MOUNT_PATH | cut -f 1
+}
+
+# Brings down the current PVC utilization to FLUENT_BIT_RECLAIM_TARGET_PERCENT by deleting the oldest compressed log files.
+# Makes the assumption that rate of deletion will be never be lower than rate of accumulation.
+reclaim() {
+    # This must be a list of file paths in sorted order.
+    local total_kib=$(($FLUENT_BIT_PVC_CAPACITY * megabyte))
+    local threshold_kib=$(((total_kib * $FLUENT_BIT_RECLAIM_TRIGGER_PERCENT)/100))
+    local used=$(used_kib)
+    if [ "$used" -lt "$threshold_kib" ]; then
+        debug "Used kilobytes ($used) less than the rotation threshold ($threshold_kib)"
+        return 0
+    fi
+    local target_kib=$(((total_kib * $FLUENT_BIT_RECLAIM_TARGET_PERCENT)/100))
+    log "Reclaiming oldest log files (Used: $total_kib, Threshold: $threshold_kib, Target: $target_kib)"
+    # Search all log folders recursively and return results from oldest to youngest.
+    local compressed=$(find . -mindepth 2 -regextype posix-extended -regex "\.\/.*\.$LOG_EXT$" -printf "%T@ %P\n")
+    compressed="$(echo "$compressed" | sort -n | grep -oE '[^ ]*$')"
+
+    local reclaimed=0
+    for file in $compressed; do
+        if [ "$(used_kib)" -gt "$target_kib" ]; then
+            bytes=$(stat "$file" -c=%s | sed 's/=//g')
+            ((reclaimed+=bytes))
+            rm "$file"
+        else
+            break
+        fi
+    done
+    local end_kib=$(used_kib)
+    if [ "$reclaimed" -gt 0 ]; then
+        utilization=$(((end_kib * 100)/total_kib))
+        log "Reclaimed a total of $((reclaimed/gigabyte))GiB ($((reclaimed/megabyte))MiB). Total(MiB): $((total_kib/kilobyte)) , Used: $((end_kib/kilobyte)) (${utilization}%)"
+    fi
+    set +
+}
+
+# This function assumes the '-%s' dateformat will be applied. It transforms any files in the '$MOUNT_PATH'
+# directory in the '<logname>.log-<epoch>.gz' format to '<logname>-<utc>.log.gz'.
+# Furthermore the logrotate epoch is also transformed to a last modified epoch.
+epoch_to_utc() {
+  suffix=".log"
+  name=$1
+  match=$(echo "$name" | grep -oE "\-[0-9]+\.$LOG_EXT$")
+  if [ $? -eq 0 ]; then
+    epoch="$(stat $name -c %Y)"
+    utc=$(date --utc +%FT%TZ -d "@$epoch")
+    original=${name%$match}
+    dest="${original%$suffix}-${utc}${suffix}.$LOG_EXT"
+    mv "$name" "$dest"
+  fi
+}
+
+# Catch any 'orphans' -- those .log files that are from an old/restarted container and no longer
+# will receive any new appends. This does not work for restarted pods in a deployment because
+# instead of just the container tag changing, a five character string is also appended to the pod name.
+# So even if the container id is different and the log file is older, we can't differentiate between
+# a restarted pod and simply a longer lived pod.
+orphans() {
```

Comment:
I think it does, but just to confirm, this method compresses the `.log` files left without being rolled on a compressed file, right? If so, good point.

Reply:
Thats correct. If a pod restarts it is inevitable that there will be some remaining logs left over (and will never receive any future appends) so this tries to catch those and compress them.